mergekit-mega: compound merging using multiple yaml documents in a single merge config

nyxkrage opened the pull request:
This adds the a new script `mergekit-mega` that takes a yaml file with multiple merge configs specified as individual yaml documents and merges them in the proper order depending on interdependencies.

A couple of quirks that should probably be ironed out before merging
1. ~~Currently there's no special handling for circular dependencies, and I'm unsure how it would break.~~
2. ~~I dont know if the `out_path` in the yaml file is ideal, or if it should instead be a `name`, and the out_dir gets specified on the command line.~~
3. ~~The huggingface repo regex probably doesn't cover all the edge cases in its current form~~
4. Unsure if just skipping over existing output directories should be the default behavior or if it should just abort, and have --force and --skip be 2 flags.

cg123 commented:
This is great! I'll look this over soon and give you some comments - think there are a few minor tweaks that would clean things up a bit. Excited to get this in.

You can ignore `pre-commit` complaining about the multi-document yaml file by the way - I'll change that setting.

nyxkrage commented:
Add a basic check for circular dependencies and by henky's suggestion decided on using names for the individual merges and then specifying the working directory/out_path as a cli arg.
Using names also means that the regex for checking if its a local model or a huggingface repo is no longer needed

cg123 commented on mergekit/scripts/megamerge.py:
```diff
@@ -0,0 +1,143 @@
+                            model_lora = src["model"].split("+")
+                            # name must not have a slash to avoid path traversal
+                            # therefore, we can use it to check if its a merge from the config
+                            if "/" not in model_lora[0]:
```
It's a nice touch that you're allowing previous merges plus a LoRA applied - that didn't occur to me. Might be nice to pull this into a helper function instead of duplicating the logic below though.

nyxkrage replied:
Yep, good idea, done now

cg123 reviewed (commented):
Looks pretty much good to me! I had one nitpick, and it looks like `isort` wants a change or two. Those aside I'm happy to merge this in.

When #67 makes its way in there's infrastructure to do exactly the kind of graph wrangling you wrote in here, which could make things cleaner - but that's a ways off.

Thanks for the PR!

nyxkrage commented:
Alright, added some docstrings mostly to please pylint ran isort and moved the part about checking if its a dependent merge or not out to a separate function, so should be good to merge!

nyxkrage commented:
Actually just realized that you can't use other models from the merge as the base model! 😅

cg123 commented:
Looks great! Thanks for putting this together.
